File: front/src/app/core/marca.service.ts
import { isPlatformBrowser } from '@angular/common';
import { HttpClient } from '@angular/common/http';
import { Injectable, PLATFORM_ID, computed, inject, signal } from '@angular/core';
import { environment } from '../../environments/environment';

/** Textos + imagens de marca (white-label). Espelha MarcaResponse no back. */
export interface Marca {
  nomePlataforma: string;
  loginTitulo: string;
  loginSubtitulo: string;
  /** URL (assinada) da logomarca; null → usa a logo padrão (SVG). */
  logoUrl: string | null;
  /** URL (assinada) da imagem de fundo do login; null → só o azul. */
  loginFundoUrl: string | null;
}

/**
 * Marca dinâmica (white-label): busca nome/título/subtítulo em GET /marca (público) no boot
 * e expõe como signals. Os PADRÕES abaixo são os textos originais hardcoded — cobrem o SSR, o
 * instante antes do GET e o fail-safe (o back também cai neles se a config faltar/estiver vazia).
 * Um cache em localStorage evita "flash" do texto padrão em cargas seguintes (igual ao tema).
 */
@Injectable({ providedIn: 'root' })
export class MarcaService {
  private readonly http = inject(HttpClient);
  private readonly platformId = inject(PLATFORM_ID);
  private readonly base = `${environment.apiUrl}/marca`;
  private static readonly CACHE = 'marca-plataforma';

  private static readonly PADRAO: Marca = {
    nomePlataforma: 'Portal do Paciente',
    loginTitulo: 'A gestão do cuidado começa aqui.',
    loginSubtitulo:
      'Cadastre e acompanhe exames, consultas e informações dos pacientes — ' +
      'com segurança e agilidade no dia a dia da equipe.',
    logoUrl: null,
    loginFundoUrl: null,
  };

  private readonly marca = signal<Marca>(MarcaService.PADRAO);

  readonly nomePlataforma = computed(() => this.marca().nomePlataforma);
  readonly loginTitulo = computed(() => this.marca().loginTitulo);
  readonly loginSubtitulo = computed(() => this.marca().loginSubtitulo);
  /** URL da logomarca configurada (null → os componentes usam o SVG padrão). */
  readonly logoUrl = computed(() => this.marca().logoUrl);
  /** Valor CSS do fundo do login: `url("...")` ou null (para [style.--login-fundo]). */
  readonly loginFundoCss = computed(() => {
    const u = this.marca().loginFundoUrl;
    return u ? `url("${u}")` : null;
  });

  /** A logo assinada falhou/expirou ao carregar: descarta para os componentes caírem no SVG. */
  descartarLogo(): void {
    this.marca.update((m) => ({ ...m, logoUrl: null }));
    this.aplicarFavicon(null);
  }

  /** Boot: aplica a marca em cache (sem flash) e busca a atual do servidor. */
  carregar(): void {
    if (!isPlatformBrowser(this.platformId)) return;
    const cache = this.lerCache();
    if (cache) {
      const c = this.normalizar(cache);
      this.marca.set(c);
      this.aplicarFavicon(c.logoUrl);
    }
    this.http.get<Marca>(this.base).subscribe({
      next: (m) => {
        const resolvida = this.normalizar(m);
        this.marca.set(resolvida);
        this.salvarCache(resolvida);
        this.aplicarFavicon(resolvida.logoUrl);
      },
      error: () => {
        /* mantém o cache/padrão */
      },
    });
  }

  /** Campo de texto vazio → cai no padrão; logo vazia → null (usa o SVG). */
  private normalizar(m: Partial<Marca> | null): Marca {
    return {
      nomePlataforma: m?.nomePlataforma?.trim() || MarcaService.PADRAO.nomePlataforma,
      loginTitulo: m?.loginTitulo?.trim() || MarcaService.PADRAO.loginTitulo,
      loginSubtitulo: m?.loginSubtitulo?.trim() || MarcaService.PADRAO.loginSubtitulo,
      logoUrl: m?.logoUrl?.trim() || null,
      loginFundoUrl: m?.loginFundoUrl?.trim() || null,
    };
  }

  /** Troca o favicon da aba pela logo configurada; sem logo, volta ao favicon padrão. */
  private aplicarFavicon(logoUrl: string | null): void {
    const link = document.querySelector<HTMLLinkElement>('link[rel~="icon"]');
    if (link) link.href = logoUrl || '/favicon.ico';
  }

  private lerCache(): Marca | null {
    try {
      const raw = localStorage.getItem(MarcaService.CACHE);
      return raw ? (JSON.parse(raw) as Marca) : null;
    } catch {
      return null;
    }
  }

  private salvarCache(m: Marca): void {
    try {
      localStorage.setItem(MarcaService.CACHE, JSON.stringify(m));
    } catch {
      /* localStorage indisponível: sem cache, sem problema */
    }
  }
}

File: back/src/main/java/com/example/pop/marca/MarcaService.java
package com.example.pop.marca;

import java.time.Duration;

import org.springframework.stereotype.Service;

import com.example.pop.configuracao.ChaveConfiguracao;
import com.example.pop.configuracao.ConfiguracaoService;
import com.example.pop.storage.StorageService;

/**
 * Resolve os textos de marca (white-label) das configurações, com fail-safe para os
 * valores padrão (os mesmos hardcoded de antes). Mesmo espírito do {@code TemaService}:
 * se a config faltar/quebrar/estiver vazia, cai no padrão — a tela nunca fica sem texto.
 */
@Service
public class MarcaService {

    /** Padrões = os textos originais hardcoded do login (fail-safe). */
    static final String NOME_PADRAO = "Portal do Paciente";
    static final String TITULO_PADRAO = "A gestão do cuidado começa aqui.";
    static final String SUBTITULO_PADRAO =
            "Cadastre e acompanhe exames, consultas e informações dos pacientes — "
                    + "com segurança e agilidade no dia a dia da equipe.";

    /** Validade da URL assinada da logo — longa para o cache do front não servir link morto. */
    private static final Duration VALIDADE_LOGO = Duration.ofHours(24);

    private final ConfiguracaoService configuracaoService;
    private final StorageService storageService;

    public MarcaService(ConfiguracaoService configuracaoService, StorageService storageService) {
        this.configuracaoService = configuracaoService;
        this.storageService = storageService;
    }

    /** Textos + imagens de marca atuais (com fallback ao padrão em cada campo). */
    public MarcaResponse marca() {
        return new MarcaResponse(
                nomePlataforma(),
                lerTextoOu(ChaveConfiguracao.LOGIN_TITULO, TITULO_PADRAO),
                lerTextoOu(ChaveConfiguracao.LOGIN_SUBTITULO, SUBTITULO_PADRAO),
                logoUrl(),
                urlImagem(ChaveConfiguracao.LOGIN_FUNDO));
    }

    /** URL assinada da logomarca (24h) ou null se não houver imagem/config — front cai no SVG. */
    public String logoUrl() {
        return urlImagem(ChaveConfiguracao.LOGO_PLATAFORMA);
    }

    /** Bytes da logomarca (server-side) para embutir no PDF; null se não houver/for inacessível. */
    public byte[] logoBytes() {
        String url = imagemBruta(ChaveConfiguracao.LOGO_PLATAFORMA);
        return url == null ? null : storageService.baixarBytes(url);
    }

    /** URL assinada (24h) da imagem da chave, ou null se não houver — fail-safe. */
    private String urlImagem(String chave) {
        String url = imagemBruta(chave);
        return url == null ? null : storageService.urlVisualizacao(url, VALIDADE_LOGO);
    }

    /** URL canônica salva na config de imagem (não assinada); null se ausente/vazia/erro. */
    private String imagemBruta(String chave) {
        try {
            String url = configuracaoService.lerImagem(chave);
            return url == null || url.isBlank() ? null : url;
        } catch (RuntimeException e) {
            return null; // config ausente/tipo divergente → sem imagem
        }
    }

    /** Nome da plataforma (config ou padrão) — reaproveitado pelo rodapé dos relatórios. */
    public String nomePlataforma() {
        return lerTextoOu(ChaveConfiguracao.NOME_PLATAFORMA, NOME_PADRAO);
    }

    private String lerTextoOu(String chave, String padrao) {
        try {
            String valor = configuracaoService.lerTexto(chave);
            return valor == null || valor.isBlank() ? padrao : valor;
        } catch (RuntimeException e) {
            return padrao; // config ausente (faltou migration) ou tipo divergente → padrão
        }
    }
}
